### PR TITLE
feat(ui-components): implement Tabs components

### LINKED ISSUE #47247: Implement Tabs component
## where to use

- ~~`client/src/components/Donation/duration-amount-options.tsx`~~(for a/b test)

- `client/src/templates/Challenges/classic/mobile-layout.tsx`

image

  ```tsx
  <Tabs
      activeKey={currentTab}
      defaultActiveKey={currentTab}
      id='mobile-layout'
      onSelect={this.switchTab}
      >
  	{/* omit something */}
      <TabPane
          eventKey={Tab.Editor}
          title={i18next.t('learn.editor-tabs.code')}
          {...editorTabPaneProps}
          >
          {/* omit something */}
          {editor}
      </TabPane>
      <TabPane
          eventKey={Tab.Console}
          title={i18next.t('learn.editor-tabs.tests')}
          {...editorTabPaneProps}
          >
          {testOutput}
      </TabPane>
      {/* omit something */}
  </Tabs>
  ```

### Tabs

| **Name**    | **Type**            | **Default** | **Description**                                    |
| ----------- | ------------------- | ----------- | -------------------------------------------------- |
| `activeKey` | `string` | `number` |             | Mark the Tab with a matching `eventKey` as active. |
| `animation` | `boolean`           | true        | Sets if `Fade` animation strategy                  |
| id          | `string`            |             |                                                    |
| onSelect    | `function`          |             | Callback fired when a Tab is selected              |

### Tab

| **Name**   | **Type**  | **Default** | **Description**        |
| ---------- | --------- | ----------- | ---------------------- |
| `disabled` | `boolean` |             | Use `false` to disable |
| `title`    | `node`    |             |                        |

### TabPane

| **Name**        | **Type**            | **Default** | **Description**                                              |
| --------------- | ------------------- | ----------- | ------------------------------------------------------------ |
| `eventKey`      | `string` | `number` |             | Uniquely identify the `<TabPane>` among its siblings.        |
| `title`         |                     |             |                                                              |
| `mountOnEnter`  |                     |             | Wait until the first "enter" transition to mount the tab (add it to the DOM) |
| `unmountOnExit` |                     |             | Unmount the tab (remove it from the DOM) when it is no longer visible |

## Screenshots

| Status     | Light theme                                                  | Dark theme                                                   |
| ---------- | ------------------------------------------------------------ | ------------------------------------------------------------ |
| selected   | image | image |
| unselected | image | image |

### Tabs

- Themes:
  - Light
  - Dark

### Tab

- Font-family: `Roboto Mono, monospace`
- States:
  - disabled
  - selected

### TabPane

- display:
  - block
  - hidden

### Tab

- border: `1px`
- border-color: `transparent`
- border-bottom-color: `#ddd`
- margin-right: `2px`
- padding: `5px 10px`
- font-size: `0.8em`
- font-color: `--foreground-secondary: var(--gray05)`

| Status     | Spacing and typography                                       |
| ---------- | ------------------------------------------------------------ |
| selected   | `border-top-color: #ddd`<br>`border-right-color: #ddd`<br>`border-left-color: #ddd`<br>`background-color: --background-quaternary` |
| unselected | `border-bottom-color: #ddd`                                  |

### TabPane

| Status     | Spacing and typography |
| ---------- | ---------------------- |
| selected   | `display: block`       |
| unselected | `display: none`        |

## rough delivery plan

- [ ] basic fCC Tailwind style Uncontrolled Tabs - #47313
- [ ] add `disable` prop
- [ ] add Fade `animation`, and could set `false`
- [ ] basic controlled Tabs, add `onSelect`
- [ ] `mountOnEnter` and `unmountOnExit`
- [ ] implement TabPane (this part of the details still need to continue to discuss)

### PR DESCRIPTION
**Pull request by masterX89**
Checklist:

- [x] I have read [freeCodeCamp's contribution guidelines](https://contribute.freecodecamp.org).
- [x] My pull request has a descriptive title (not a vague title like `Update index.md`)
- [x] My pull request targets the `main` branch of freeCodeCamp.
- [x] I have tested these changes either locally on my machine, or GitPod.

part #47247, 
plan step one

closes

I would like to know if I am allowed to use `@headlessui/react` After reading #47043, if I can use `Transition` in `@headlessui/react`, the implementation of animation will be simpler and more elegant

### CODE REVIEW

**Comment by ieahleen** [comment]
> part Closes #47247, plan step one

if it's only partly, don't use the Closes keyword, or the PR is going to close the issue when merged

**Comment by ahmadabdolsaheb** [comment]
Thank you for for your patience with the review. 

This is a great start and here are a few points I would like to discuss about this implementation.

1. Yes, `@headlessui/react` simplifies the development a lot. If you decide to go down that route, please make sure  `Tabs` and `TabPane` still receive the same props and have the same structure when used.
2. If you decide develop the component from scratch, I highly recommend to also look into how the current version of [`Tabs` ](https://github.com/react-bootstrap/react-bootstrap/blob/master/src/Tab.tsx) and its related components are developed and how the state is managed.

4. The current usage of `Tabs` and `TabPane` could be simplified in learn so we could spend less time creating the `Tabs` related components. For instance, `TabPane` could be replaced with `Tab` and `onSelection` could be completely removed.
5. I would appreciate it if you could do the research for the  `mountOnEnter` and ` unmountOnExit` and learn about performance benefits and see how we could implement them.

6. Animation is not needed at this point. But we will include it across our components if necessary in the next iterations.

Let me know if the above makes sense to you.

**Comment by raisedadead** [comment]
Hi, 

We are closing old issues/pull requests that we think are stale. If you still need help or are fixing this, please open a new issue/pull request.

Thanks for your understanding. 

Happy contributing.

**Comment by ahmadabdolsaheb** [comment]
@Sboonny, here is the tabs component for your consideration.

**Comment by masterX89** [comment]
I'm sorry for not responding to the PR here for being busy last year.

After China has fully released the lockdown, I will discuss this PR with members of the freeCodeCamp community in Shanghai this weekend and try to advance it. (My train of thought is a bit stucked😂)

**Comment by Sboonny** [comment]
Welcome back, I have discussed the abstraction in tabpane with Ahmad, instead of `tabPane` accepting children and tabs mapping over the tabpane and creating multiple content and buttons.

for example, in mobile layout this how code look like

https://github.com/freeCodeCamp/freeCodeCamp/blob/0f922b86522cba6ec85e4169145e1faaa92cfbb3/client/src/templates/Challenges/classic/mobile-layout.tsx#L87-L125

and this will create tabTrigger and tabContent

```ts
   <TabTrigger role='tab' value={Tab.Instructions}>
          {i18next.t('learn.editor-tabs.instructions')}
    </TabTrigger>
    <TabTrigger role='tab' value={Tab.Editor}>
          {i18next.t('learn.editor-tabs.code')}
     </TabTrigger>

      <TabContent value={Tab.Instructions} {...props}>
        {instructions}
      </TabContent>
      <TabContent value={Tab.Editor} {...props}>
        {editor}
      </TabContent>
```

Instead of logic that creates this, we want to abandon that destructure and restructuring of the elements, and just create a direct `TabTrigger` and `TabContent` that will show and hide the content when the value is the same.

**Comment by raisedadead** [comment]
@masterX89 are you still working on this? @ahmadabdolsaheb @Sboonny should you guys take this up instead?

**Comment by masterX89** [comment]
Sorry, I'm busy with my work again and I probably won't be free until early July.

May need help from others to complete this component.

**Comment by Sboonny** [comment]
No worries, you have done most of the work anyone, I will add the finishing touches at the end of the day.

**Comment by Sboonny** [comment]
storybook isn't reading the value of the variables names in the CSS, this is applying to the main branch, too.

I can't duplicate the styles unless this is fixed, so I am putting this on hiatus until I solve the storybook issue.

**Comment by ahmadabdolsaheb** [comment]
@Sboonny thank you for working on this pr. Let me know if you need my help to finalize it.

**Comment by Sboonny** [comment]
No need, if you are happy with the current implementation, I will sort the conflicts and we merge this.

File: tools/ui-components/src/tabs/tabs.tsx [hunk 1]
```diff
@@ -0,0 +1,31 @@
+import React from 'react';
+import { Root, List, Trigger, Content } from '@radix-ui/react-tabs';
+
+const buttonClassNames =
+  'flex-1 block relative px-2.5 py-[5px] mr-[2px] border-1 border-solid border-background-tertiary border-b-transparent text-sm text-foreground-secondary aria-selected:border-0 aria-selected:font-black aria-selected:bg-foreground-quaternary aria-selected:text-background-secondary';
```
Comment (Sboonny) [inline tools/ui-components/src/tabs/tabs.tsx:5]:
```suggestion
  'flex-1 block relative px-2.5 py-[5px] text-sm text-foreground-secondary aria-selected:border-0 aria-selected:font-black aria-selected:bg-foreground-quaternary aria-selected:text-background-secondary';
```

**Review by ahmaxed — APPROVED** [review]

**Review by Sembauke — APPROVED** [review]

**Pull request merged**